app/src/ui/dialog/dialog.tsx

```diff
@@ -143,9 +143,56 @@ export class Dialog extends React.Component<IDialogProps, IDialogState> {
   private disableClickDismissalTimeoutId: number | null = null
   private disableClickDismissal = false
 
+  /**
+   * Resize observer used for tracking width changes and
+   * refreshing the internal codemirror instance when
+   * they occur
+   */
+  private readonly resizeObserver: ResizeObserver
+  private resizeDebounceId: number | null = null
+
   public constructor(props: IDialogProps) {
     super(props)
     this.state = { isAppearing: true }
+
+    // Observe size changes and let codemirror know
+    // when it needs to refresh.
+    this.resizeObserver = new ResizeObserver(this.scheduleResizeEvent)
+  }
+
+  private scheduleResizeEvent = () => {
+    if (this.resizeDebounceId !== null) {
+      cancelAnimationFrame(this.resizeDebounceId)
+      this.resizeDebounceId = null
+    }
+    this.resizeDebounceId = requestAnimationFrame(this.onResized)
+  }
+
+  /**
+   * Attempt to ensure that the entire dialog is always visible. Chromium
+   * takes care of positioning the dialog when we initially show it but
+   * subsequent resizes of either the dialog (such as when switching tabs
+   * in the preferences dialog) or the Window doesn't affect positioning.
+   */
+  private onResized = () => {
+    if (!this.dialogElement) {
+      return
+    }
+
+    const { offsetTop, offsetHeight } = this.dialogElement
+
+    // Not much we can do if the dialog is bigger than the window
+    if (offsetHeight > window.innerHeight - titleBarHeight) {
+      return
+    }
+
+    const padding = 10
+    const overflow = offsetTop + offsetHeight + padding - window.innerHeight
+
+    if (overflow > 0) {
+      const top = Math.max(titleBarHeight, offsetTop - overflow)
+      this.dialogElement.style.top = `${top}px`
+    }
```

Comment:
i would really like to solve this in our scss if possible. @niik is there something preventing us from writing roughly equivalent logic into our base dialog styles?

Reply:
> is there something preventing us from writing roughly equivalent logic into our base dialog styles?

Yeah, this is something we talked about at when we initially implemented the dialogs. The chromium logic takes care of centering the dialog when it's first opened but doesn't move the dialog after that. That's the behavior we want. If we were to use CSS to position all dialogs at the vertical midpoint they'd start jumping around when switching between tabs in dialogs or when the contents of the dialog changes for other reasons (error messages dynamically appearing/disappearing for example).

The problem we're running into here is when a dialog can have wildly different heights depending on its configuration which is the case for the preferences dialog where the entry tab is very short:

<img width="617" alt="image" src="https://user-images.githubusercontent.com/634063/75876726-83b21900-5e16-11ea-84eb-bd68741520ba.png">

And the advanced tab is much much taller

<img width="614" alt="image" src="https://user-images.githubusercontent.com/634063/75876746-90cf0800-5e16-11ea-9332-4ac922d63fe4.png">

I see this change in this PR as a safe-guard to ensure that we don't ever end up in scenarios where the dialog ends up extending beyond the bounds of the window. A much better solution here would be to rebalance the tabs in the preferences dialog so that each tab is more uniform in height. We could for example add the "save/cancel" footer to the accounts and theme tabs and we could consider extracting the "show a confirmation dialog when..." section of the advanced tab into its own "Prompts" tab.

TL;DR; I don't see this as a solution in any way shape or form. Our dialogs aren't intended to differ wildly in height and we should ensure that the preferences dialog doesn't do that either. It is, however, a safe-guard.